### DISCUSSION TITLE
FROM table JOIN read_parquet(table.uri)

### DISCUSSION THREAD

**Discussion by liquidcarbon** [duckdb/duckdb#13750] [Ideas]
Imagine that you have a table that contains fields pointing to locations of other datasets.

It's easy enough to do two separate queries, first to retrieve a list of URIs, second to read those URIs.

What if you could use those fields to construct the arguments to JOINs directly?

```python
con = duckdb.connect()
con.execute("CREATE TABLE a (id INTEGER, uri TEXT)")
con.execute("INSERT INTO a VALUES (1, 'path/to/file1.parquet'), (2, 's3://path/to/file2.parquet')")
result = con.execute("SELECT * FROM a JOIN read_parquet(a.file_path) USING (id)").df()
```
`BinderException: Binder Error: Table function "read_parquet" does not support lateral join column parameters - cannot use column "a.file_path" in this context`

What would it take to make this possible?

**Comment by Maxxen** [comment]
This is not generally possible to do in SQL as all the data types and schemas must be known at "bind time", before the query gets executed so that DuckDB can setup, typecheck, allocate, optimize and otherwise prepare and "plan" the pipeline. 

There are some table functions that support these "lateral join parameters", i.e. so called "table-in-out" functions, but those require that the resulting output has a fixed schema, (e.g the "range" table function supports this)

The solution is to simply run two queries.

**Comment by liquidcarbon** [comment]
Well... aren't some queries actually more than one query behind the scenes?

Here, one could imagine the first query reading just `a` so that the argument in `read_parquet(a.file_path)` could be replaced with a list of paths, that's all that needs to happen (probably an oversimplification).

What's a "table-in-out" example?

**Reply by Maxxen** [reply]
> aren't some queries actually more than one query behind the scenes?

What do you mean? Are you thinking of subqueries? Subqueries are not executed separately but form a part of the bigger plan.

>  the argument in read_parquet(a.file_path) could be replaced with a list of paths

What would `SELECT * FROM  read_parquet(a.file_path)` return? There is no way to know what the resulting schema would be or how to plan the query without first executing and finishing the query resolving `a`. In other words, run two queries one after the other. Which is not something any database I know of really does.

Mark went into [more detail in this thread](https://github.com/duckdb/duckdb/discussions/12505#discussioncomment-9754240
)

> What's a "table-in-out" example?

On DuckDB main, the `range` table function is a table-in-out function, meaning that its parameters don't have to be constants. You can use it in a lateral join, passing a column from another "adjacent" relation in the same FROM clause into it as an argument.

```sql
CREATE TABLE t1 AS SELECT * FROM (VALUES (1), (2), (3)) t(i);
SELECT j FROM t1, range(0, i) as r(j);
┌───────┐
│   j   │
│ int64 │
├───────┤
│     0 │
│     1 │
│     0 │
│     0 │
│     1 │
│     2 │
└───────┘
```

**Reply by liquidcarbon** [reply]
1) I don't get how this works... how do you end up with 0, 1, 0, 0, 1, 2?

2) subqueries, unions, whatever - yes, one query plan is made, but it's ran stepwise 

3) `SELECT * FROM  read_parquet(a.file_path)` wouldn't make sense without `a`; `FROM a, read_parquet(a.path)` becomes just like the example in #12505 `SELECT * FROM tbl, read_csv(tbl.file);` 

> DuckDB supports table functions whose output schema is dependent on the input. For example, DuckDB supports the read_csv function

I suppose I don't understand query planning.  In my mind, the query may be pre-processed into

```sql
FROM a
JOIN read_parquet([path1, path2]) USING (id)
```
by running a simple query on `a` first, to populate the paths list.  It's like syntactic sugar, if you will.


DuckDB has done many things that never existed in other databases, that's why we love it!

**Comment by liquidcarbon** [comment]
This is related to https://github.com/duckdb/duckdb/discussions/8487 - still looking for clever ways to avoid scanning a very large collection of files.

The proposed pattern could be also be helpful when you need to scan multiple files scattered all over (some local, some in clouds).

**Comment by liquidcarbon** [comment]
This [new feature](https://github.com/duckdb/duckdb/pull/13084) in v1.1 does what I was suggesting here - feel free to close if you agree.
```
SET VARIABLE list_of_files = (SELECT LIST(file) FROM csv_files);
SELECT * FROM read_csv(getvariable('list_of_files'), filename := true);
```
